### PR TITLE
Elevator "State Machine"

### PR DESCRIPTION
**Pull request by JackyChiu**
This PR adds simulation classes for the physical elevator components and the code to handle the elevator messages telling the elevator to move up/down, and close/open it's doors.

### CODE REVIEW

File: src/elevator/Elevator.java [hunk 1]
```diff
@@ -4,91 +4,122 @@
...
+            }
+    }
+
+    /*
+     * handleElevatorMessage handles commands for the elevator's physical components.
+     */
+    private void handleElevatorMessage(ElevatorMessage m) {
+        switch (m.getMessageType()) {
+            case STOP:
+                assert(state == State.MOVING_UP || state == State.MOVING_DOWN);
```
Comment (JackyChiu) [inline src/elevator/Elevator.java]: These assertions are going to help up debug ensure that our elevator state machine is following the path we expect. I'm not sure how we can handle out of order messages.

File: src/elevator/Elevator.java [hunk 2]
```diff
@@ -4,91 +4,122 @@
...
+
+    /*
+     * handleMessage runs the corresponding action for the message type.
+     */
+    private void handleMessage(Message m) {
+        // State machine switch
+        if (m instanceof ElevatorMessage) {
+            // scheduler tells the elevator to move, stop, open or close.
+            handleElevatorMessage((ElevatorMessage) m);
+            // } else if (m instanceof FloorMessage) {
```
Comment (vikrambombhi) [inline src/elevator/Elevator.java]: Can u remove this `else if`

File: src/elevator/Elevator.java (hunk 1 above)
Comment (vikrambombhi) [reply src/elevator/Elevator.java]: Assume a ideal network

File: src/elevator/Elevator.java [hunk 3]
```diff
@@ -4,91 +4,122 @@
...
+                assert(state == State.STOPPED_DOORS_CLOSED);
+                door.open();
+                state = State.STOPPED_DOORS_OPENED;
+                break;
+
+            case CLOSEDOOR:
+                assert(state == State.STOPPED_DOORS_OPENED);
+                door.close();
+                state = State.STOPPED_DOORS_CLOSED;
+                break;
```
Comment (vikrambombhi) [inline src/elevator/Elevator.java]: Should probably have a default case for invalid messages

File: src/elevator/Elevator.java (hunk 1 above)
Comment (JackyChiu) [reply src/elevator/Elevator.java]: we're already using udp so I guess this is the correct way forward

File: src/elevator/Elevator.java (hunk 3 above)
Comment (JackyChiu) [reply src/elevator/Elevator.java]: What should the default case do? Blow up just for the sake of being caught by us?

**Pull request merged**